### DISCUSSION TITLE
BAC0 device

### DISCUSSION THREAD

**Discussion by Kabilan001** [ChristianTremblay/BAC0#452] [General]
@ChristianTremblay big fan of your work. I have a doubt regarding the usage of BAC0

This is my query to collect all raw data over BACnet. I setup an emulator to test it. The present value of the emulator is set to change every minute. However, the BAC0 device points output does not get updated with the latest values after the first read. When I disconnect and re connect the controller, the new value gets read but does not updated after that. Am I missing something?

```python
import BAC0
import bacpypes
import threading
from datetime import datetime
import logging
import pymongo
import queue
import time

logging.basicConfig(filename='data_collection.log', level=logging.WARNING, format='%(asctime)s - %(levelname)s - %(message)s')

bacpypes.settings.settings['route_aware'] = False

network_ip = "<IP>"
bacnet = BAC0.connect(network_ip)

bacnet.whois(global_broadcast=True)
time.sleep(5)
print(bacnet.devices)



data_queue = queue.Queue()

NUM_FETCH_THREADS = len(bacnet.devices)

def collect_data(device):
    device_ip = device[2]
    device_id = device[3]
    mycontroller = BAC0.device(device_ip, device_id, bacnet,poll=1,clear_history_on_save=True,history_size=1)

    while True:
        points = mycontroller.points
        print("Read the new points")
        start_time = time.time()
        data = []
        for point in points:
            properties = point.bacnet_properties
            properties.pop("eventTimeStamps", None)
            properties.pop("priorityArray", None)
            properties.pop("changeOfStateTime", None)
            properties.pop("timeOfStateCountReset", None)
            properties.pop("timeOfActiveTimeReset", None)
            data.append(properties)
        
        raw_payload_to_upload = {
            "DeviceIP": str(device_ip),
            "DeviceID:": str(device_id),
            "Timestamp": int(time.time()),
            "data": data,
        }
        points = None
        print("Raw Data pushed to: %s ", raw_payload_to_upload)
        # Record end time
        end_time = time.time()
        # Calculate elapsed time
        elapsed_time = end_time - start_time
        print("Time taken to run the script: %s seconds", elapsed_time)
        time.sleep(10)


database_thread = threading.Thread(target=insert_into_database)
database_thread.daemon = True
database_thread.start()

fetch_threads = []
for device in bacnet.devices:
    for _ in range(NUM_FETCH_THREADS):
        t = threading.Thread(target=collect_data, args=(device,))
        fetch_threads.append(t)
        t.start()

for t in fetch_threads:
    t.join()

data_queue.join()
```

**Comment by ChristianTremblay** [comment]
Each time you `time.sleep(10)` you kill the app...

The role of a `BAC0.device` is to be kept running to poll data. 1 second will probably too fast. 

I would propose you trust the `BAC0.device` and use custom object list  to target the point you really need. Each device already run thread to collect data.

Use a history_size small but big enough for your use case.

For the other part of your script processing the data, use `device['point'].lastValue` (or use `device['point'].history` pandas dataframe)

**Comment by ChristianTremblay** [comment]
Better speed will be possible with the async version.... but 1 second, to me, is too fast

**Reply by Kabilan001** [reply]
Noted @ChristianTremblay . Thank you for your reply. Instead of fetching every data point 1 by 1 with "device['point'].lastValue", is there a way to get all the updated data points in my custom object list?

**Reply by ChristianTremblay** [reply]
lastValue will not fetch on the network but take the value from an internal ._history variable that is present for each point. I guess you could merge all the dataframe at once and retreive that...

```python
for each device.points:
    #do something with each.lastValue
```
